File: src/cs/Samples/Samples/Extensions.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Windows.Forms;

namespace ApiSamples
{
    static class ControlExtensions
    {
        public static void Do<TControl>(this TControl control, Action<TControl> action)
          where TControl : Control
        {
            if (control.InvokeRequired)
            {
                //control.Invoke(action, control);
                control.BeginInvoke(action, control);
            }
            else
            {
                action(control);
            }
        }
    }
}

File: General/OpenSave/vb/OpenSave/Extensions.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms

Friend Module ControlExtensions
    <System.Runtime.CompilerServices.Extension> _
    Public Sub [Do](Of TControl As Control)(ByVal control As TControl, ByVal action As Action(Of TControl))
        If control.InvokeRequired Then
            control.BeginInvoke(action, control)
        Else
            action(control)
        End If
    End Sub
End Module
